Significant overhaul to container+dependency management

jmchilton opened the pull request:
*Overview*

- Migrate "manage dependencies" UI to APIs and VueJS.
  - Add new `summarize_toolbox` API endpoint to mimic information produced by former grid.
  - The tool and requirement views are in the same grid now - depending on what you are viewing there is a "collapse tools" / "expand tools" button to swap to the other view.
  - Unused environments grid migrated as its own component.
  - By default information displayed information is more compact but every row can be expanded and many elements display a lot more information on mouse over.
- Implement fallback container resolvers (describe below).
- Implement RESTful API around container resolvers - describe both how resolvers are configured and the resolution of particular tools (or all tools). Add API endpoint for building containers as well.
- Implement new VueJS table describing container resolution independent of dependencies. Allow mixed mode (show whatever container first resolved) as well as filtering by Docker/Singularity.
  - This information about containers is much more correct than what the manage dependencies grid would display - which was independent of tool IDs and based only on requirements, would not query quay.io for information, etc..
  - Since the grid is restricted to containers, embedded information is richer - displays information about the container resolver, a link to the container, correct table header, shell used by the container, etc..

*Why Fallback Container Resolvers?*

tl;dr - we can write better config files and more useful UIs if we specify default containers this way.

Default containers can be specified in the job_conf but Galaxy doesn't really have the ability to dispatch or reason about such configuration until it is time to run the job. galaxy.tool_util has no clue about job configuration structure and job configuration structure can be (and often is) completely dynamic. As a result, such configuration should be thought of procedurally specified - configuration via container resolvers is effectively more explicitly annotated.

This has two upshots for my current cloud dev push:

- Containers specified this way can reasoned about in terms of the dependency magnagement API so we can provide an interface for seeing what tools will map to job (by default anyway - same stipulation applies to mulled containers which can also be overridden). I will utilize to make #8741.
- While I think job_conf should ultimately allow overriding the container, I think the way this works out in most setups - including the CloudVE setup (https://github.com/galaxyproject/galaxy-helm/blob/master/galaxy/values.yaml#L201) the container is more realistically a global configuration setting. As such it is not naturally specified job_conf, awkardness results, and it would be much more cleanly specified in the container configuration - a global configuration.

*Existing UI in 19.09*

<img width="1374" alt="Screen Shot 2019-10-02 at 12 40 18 PM" src="https://user-images.githubusercontent.com/216771/66063719-f6d6fc80-e511-11e9-990d-b1f9ae1d6d1c.png">

*New UI*

<img width="1412" alt="Screen Shot 2019-11-07 at 3 18 52 PM" src="https://user-images.githubusercontent.com/216771/68424673-7a6bb500-0172-11ea-9544-9df4ae594c02.png">

---

<img width="1415" alt="Screen Shot 2019-11-07 at 3 19 05 PM" src="https://user-images.githubusercontent.com/216771/68424672-7a6bb500-0172-11ea-8a1b-93c69c5f5839.png">

---

<img width="1404" alt="Screen Shot 2019-11-07 at 3 19 21 PM" src="https://user-images.githubusercontent.com/216771/68424671-7a6bb500-0172-11ea-930e-389cea078493.png">

---

<img width="1398" alt="Screen Shot 2019-11-07 at 3 19 33 PM" src="https://user-images.githubusercontent.com/216771/68424670-79d31e80-0172-11ea-94a8-b5ce11dcbeea.png">

---

<img width="1410" alt="Screen Shot 2019-11-07 at 3 19 41 PM" src="https://user-images.githubusercontent.com/216771/68424669-79d31e80-0172-11ea-923c-ce6390f212e9.png">

---

<img width="1396" alt="Screen Shot 2019-11-07 at 3 20 01 PM" src="https://user-images.githubusercontent.com/216771/68424668-79d31e80-0172-11ea-8393-5b9c1c23e329.png">

---

<img width="1406" alt="Screen Shot 2019-11-07 at 3 20 11 PM" src="https://user-images.githubusercontent.com/216771/68424667-79d31e80-0172-11ea-9bfd-f9d76c9ab6da.png">

---

<img width="1412" alt="Screen Shot 2019-11-07 at 3 20 27 PM" src="https://user-images.githubusercontent.com/216771/68424666-79d31e80-0172-11ea-8447-7e9b38324ece.png">

---

<img width="1416" alt="Screen Shot 2019-11-07 at 3 21 47 PM" src="https://user-images.githubusercontent.com/216771/68424665-79d31e80-0172-11ea-962c-f3af9d4c4e1f.png">


*Newer UI (earlier iterations)*

<img width="975" alt="Screen Shot 2019-10-11 at 4 36 58 PM" src="https://user-images.githubusercontent.com/216771/66683424-669d6380-ec45-11e9-9fba-715a60741d15.png">

---

<img width="966" alt="Screen Shot 2019-10-11 at 4 34 39 PM" src="https://user-images.githubusercontent.com/216771/66683345-3ce43c80-ec45-11e9-9f74-896e667bb8c4.png">

---

<img width="978" alt="Screen Shot 2019-10-11 at 4 34 25 PM" src="https://user-images.githubusercontent.com/216771/66683346-3ce43c80-ec45-11e9-928c-2c15d348ad8e.png">

---

<img width="953" alt="Screen Shot 2019-10-11 at 4 33 47 PM" src="https://user-images.githubusercontent.com/216771/66683347-3ce43c80-ec45-11e9-99ae-6874f1672f97.png">

martenson commented:
@jmchilton Could you rebase this please so we get a clean diff?

mvdbeek commented:
> Default containers can be specified in the job_conf but Galaxy doesn't really have the ability to dispatch or reason about such configuration until it is time to run the job. galaxy.tool_util has no clue about job configuration structure and job configuration structure can be (and often is) completely dynamic. As a result, such configuration should be thought of procedurally specified - configuration via container resolvers is effectively more explicitly annotated.

That is awesome, I agree completely!



> * While I think job_conf should ultimately allow overriding the container, I think the way this works out in most setups - including the CloudVE setup (https://github.com/galaxyproject/galaxy-helm/blob/master/galaxy/values.yaml#L201) the container is more realistically a global configuration setting. As such it is not naturally specified job_conf, awkardness results, and it would be much more cleanly specified in the container configuration - a global configuration.

💯

jmchilton commented:
@mvdbeek Thanks for the note about container resolvers - glad you agree. I've been holding off a bit a push to implement per tool default containers at the resolver level but I think with you buy in I feel more comfortable adding that.

Have you played with this at all? I feel like there are a lot of next steps here but I think I'd like to undo placing containers and dependencies in the same grid (reasons outlined here https://github.com/galaxyproject/galaxy/pull/7125#issuecomment-539516275) but if you'd like to keep the mixed mode I'll reconsider things.

jmchilton committed:
Significant overhaul to container+dependency management

*Overview*

- Migrate "manage dependencies" UI to APIs and VueJS.
  - Add new `summarize_toolbox` API endpoint to mimic information produced by former grid.
  - The tool and requirement views are in the same grid now - depending on what you are viewing there is a "collapse tools" / "expand tools" button to swap to the other view.
  - Unused environments grid migrated as its own component.
  - By default information displayed information is more compact but every row can be expanded and many elements display a lot more information on mouse over.
- Implement fallback container resolvers (describe below).
- Implement RESTful API around container resolvers - describe both how resolvers are configured and the resolution of particular tools (or all tools). Add API endpoint for building containers as well.
- Implement new VueJS table describing container resolution independent of dependencies. Allow mixed mode (show whatever container first resolved) as well as filtering by Docker/Singularity.
  - This information about containers is much more correct than what the manage dependencies grid would display - which was independent of tool IDs and based only on requirements, would not query quay.io for information, etc..
  - Since the grid is restricted to containers, embedded information is richer - displays information about the container resolver, a link to the container, correct table header, shell used by the container, etc..

*Why Fallback Container Resolvers?*

tl;dr - we can write better config files and more useful UIs if we specify default containers this way.

Default containers can be specified in the job_conf but Galaxy doesn't really have the ability to dispatch or reason about such configuration until it is time to run the job. galaxy.tool_util has no clue about job configuration structure and job configuration structure can be (and often is) completely dynamic. As a result, such configuration should be thought of procedurally specified - configuration via container resolvers is effectively more explicitly annotated.

This has two upshots for my current cloud dev push:

- Containers specified this way can reasoned about in terms of the dependency magnagement API so we can provide an interface for seeing what tools will map to job (by default anyway - same stipulation applies to mulled containers which can also be overridden). I will utilize to make https://github.com/galaxyproject/galaxy/pull/8741.
- While I think job_conf should ultimately allow overriding the container, I think the way this works out in most setups - including the CloudVE setup (https://github.com/galaxyproject/galaxy-helm/blob/master/galaxy/values.yaml#L201) the container is more realistically a global configuration setting. As such it is not naturally specified job_conf, awkardness results, and it would be much more cleanly specified in the container configuration - a global configuration.

Handle empty requirements list in ResolutionDetails.vue.

Error reported here https://github.com/galaxyproject/galaxy/pull/8741#issuecomment-550253702.

jmchilton committed:
Implement container ResolutionCache for repeated computation...

... when resolving across a toolbox. Start with caching the result of 'docker images'.

jmchilton committed:
Container resolution - save some quay requests by caching list of repos per namespace.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/container_resolution.py:
```diff
@@ -0,0 +1,121 @@
+"""
+API operations allowing clients to manage tool dependencies.
```
```suggestion
API operations allowing clients to manage container resolution.
```

mvdbeek committed:
Speed up dependency view

Profiling before:
```
*** PROFILER RESULTS ***
summarize_toolbox (/Users/mvandenb/src/galaxy/lib/galaxy/web/framework/decorators.py:228)
function called 1 times

         42850785 function calls (40691205 primitive calls) in 63.276 seconds

   Ordered by: cumulative time, internal time, call count
   List reduced from 279 to 40 due to restriction <40>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   63.276   63.276 decorators.py:228(decorator)
        1    0.000    0.000   63.272   63.272 decorators.py:87(decorator)
        1    0.000    0.000   63.272   63.272 tool_dependencies.py:189(summarize_toolbox)
        1    0.056    0.056   63.272   63.272 views.py:228(summarize_requirements)
      277    7.137    0.026   61.665    0.223 views.py:211(tool_ids_by_requirements)
   716598    2.004    0.000   26.227    0.000 requirements.py:127(__eq__)
  1586883    3.859    0.000   24.838    0.000 oset.py:8(__init__)
  1428437    5.199    0.000   20.979    0.000 _collections_abc.py:604(__ior__)
   869226    1.421    0.000   20.939    0.000 __init__.py:1664(tool_requirements)
   716598    1.772    0.000   20.415    0.000 _collections_abc.py:474(__and__)
   869226    3.086    0.000   19.518    0.000 requirements.py:115(packages)
   716598    1.240    0.000   16.914    0.000 _collections_abc.py:465(_from_iterable)
   870285    2.999    0.000   12.723    0.000 requirements.py:99(__init__)
  1677185    5.694    0.000    8.816    0.000 oset.py:21(add)
   869778    1.800    0.000    7.336    0.000 requirements.py:143(__hash__)
  1413485    3.166    0.000    6.964    0.000 _collections_abc.py:477(<genexpr>)
   869778    3.816    0.000    5.332    0.000 requirements.py:144(<listcomp>)
  5052459    4.203    0.000    4.730    0.000 requirements.py:51(__hash__)
4321842/2163046    3.081    0.000    3.828    0.000 {built-in method builtins.len}
   869226    2.847    0.000    3.734    0.000 requirements.py:117(<listcomp>)
   696887    2.075    0.000    3.291    0.000 oset.py:18(__contains__)
  5151455    2.050    0.000    2.050    0.000 oset.py:33(__iter__)
  2456292    1.339    0.000    1.918    0.000 {built-in method builtins.isinstance}
        1    0.008    0.008    1.531    1.531 views.py:241(<dictcomp>)
      276    0.023    0.000    1.496    0.005 views.py:282(get_requirements_status)
      276    0.007    0.000    1.436    0.005 views.py:52(show_dependencies)
      276    0.007    0.000    1.428    0.005 views.py:62(<dictcomp>)
      276    0.002    0.000    1.420    0.005 __init__.py:185(requirements_to_dependencies)
      276    0.083    0.000    1.418    0.005 __init__.py:198(_requirements_to_dependencies_dict)
  2154295    0.745    0.000    0.927    0.000 oset.py:15(__len__)
     2457    0.016    0.000    0.736    0.000 __init__.py:1356(debug)
     2457    0.022    0.000    0.719    0.000 __init__.py:1491(_log)
   723361    0.247    0.000    0.579    0.000 abc.py:137(__instancecheck__)
     2457    0.011    0.000    0.577    0.000 __init__.py:1516(handle)
     2457    0.011    0.000    0.565    0.000 __init__.py:1570(callHandlers)
     2457    0.018    0.000    0.554    0.000 __init__.py:881(handle)
  5052459    0.527    0.000    0.527    0.000 {built-in method builtins.hash}
     2457    0.017    0.000    0.523    0.000 __init__.py:1013(emit)
   696378    0.495    0.000    0.495    0.000 requirements.py:45(__eq__)
   711839    0.398    0.000    0.481    0.000 requirements.py:103(<listcomp>)
```

Profiling after:
```
*** PROFILER RESULTS ***
summarize_toolbox (/Users/mvandenb/src/galaxy/lib/galaxy/web/framework/decorators.py:228)
function called 1 times

         725248 function calls (707704 primitive calls) in 1.945 seconds

   Ordered by: cumulative time, internal time, call count
   List reduced from 278 to 40 due to restriction <40>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    1.945    1.945 decorators.py:228(decorator)
        1    0.000    0.000    1.941    1.941 decorators.py:87(decorator)
        1    0.000    0.000    1.941    1.941 tool_dependencies.py:189(summarize_toolbox)
        1    0.005    0.005    1.941    1.941 views.py:228(summarize_requirements)
        1    0.008    0.008    1.708    1.708 views.py:242(<dictcomp>)
      276    0.025    0.000    1.672    0.006 views.py:283(get_requirements_status)
      276    0.008    0.000    1.607    0.006 views.py:52(show_dependencies)
      276    0.008    0.000    1.599    0.006 views.py:62(<dictcomp>)
      276    0.002    0.000    1.590    0.006 __init__.py:185(requirements_to_dependencies)
      276    0.089    0.000    1.589    0.006 __init__.py:198(_requirements_to_dependencies_dict)
     2457    0.018    0.000    0.818    0.000 __init__.py:1356(debug)
     2457    0.025    0.000    0.798    0.000 __init__.py:1491(_log)
     2457    0.012    0.000    0.643    0.000 __init__.py:1516(handle)
     2457    0.011    0.000    0.630    0.000 __init__.py:1570(callHandlers)
     2457    0.019    0.000    0.619    0.000 __init__.py:881(handle)
     2457    0.018    0.000    0.586    0.000 __init__.py:1013(emit)
     2457    0.021    0.000    0.456    0.000 pydevd_io.py:30(write)
     4198    0.265    0.000    0.265    0.000 {built-in method posix.stat}
     2457    0.024    0.000    0.243    0.000 pydevd.py:2044(write)
        1    0.026    0.026    0.220    0.220 views.py:211(tool_ids_by_requirements)
     1483    0.008    0.000    0.200    0.000 galaxy_packages.py:68(resolve)
      961    0.023    0.000    0.192    0.000 conda.py:256(resolve)
     2457    0.190    0.000    0.190    0.000 {method 'write' of '_io.TextIOWrapper' objects}
     2936    0.006    0.000    0.175    0.000 genericpath.py:39(isdir)
      507    0.016    0.000    0.171    0.000 conda.py:184(resolve_all)
     2457    0.029    0.000    0.163    0.000 pydevd_net_command_factory_json.py:248(make_io_message)
      974    0.006    0.000    0.149    0.000 galaxy_packages.py:112(resolve)
      631    0.010    0.000    0.117    0.000 galaxy_packages.py:96(_galaxy_package_dep)
     2457    0.035    0.000    0.114    0.000 pydevd_net_command.py:36(__init__)
     6783    0.018    0.000    0.108    0.000 oset.py:8(__init__)
     2457    0.012    0.000    0.106    0.000 __init__.py:1476(makeRecord)
     1262    0.003    0.000    0.099    0.000 genericpath.py:16(exists)
     2586    0.007    0.000    0.094    0.000 requirements.py:127(__eq__)
     2457    0.056    0.000    0.094    0.000 __init__.py:282(__init__)
     1468    0.006    0.000    0.092    0.000 conda_util.py:318(has_env)
     6209    0.023    0.000    0.090    0.000 _collections_abc.py:604(__ior__)
      503    0.003    0.000    0.088    0.000 galaxy_packages.py:81(_find_dep_versioned)
     2457    0.006    0.000    0.077    0.000 __init__.py:858(format)
     3138    0.005    0.000    0.074    0.000 __init__.py:1664(tool_requirements)
     2586    0.006    0.000    0.073    0.000 _collections_abc.py:474(__and__)
```

mvdbeek commented:
I have now, and this is great! I'm totally fine with separating containers and dependencies, you've convinced me. One thing though is even with the new caching it takes more than an hour to list the container resolver view. Maybe we can just search for containers on demand ?
I checked the quay.io API and didn't really see an alternative for bulk requests (short of just building the list in a cronjob on our side).

Also the resolution view is a lot slower than the legacy view. The legacy view takes 3.5 seconds to load, the new view takes 51 seconds. That should be fixed by https://github.com/jmchilton/galaxy/pull/75

mvdbeek commented:
Expanding unsresolved tools without requirements also crashes in the JS console:
```
vue.runtime.esm.js?2b0e:619 [Vue warn]: Error in render: "TypeError: Cannot read property 'dependency_resolver' of undefined"

found in

---> <ResolutionDetails> at galaxy/scripts/components/admin/Dependencies/ResolutionDetails.vue
       <BTableCell>
         <BTr>
           <BTbody>
             <BTable>
               <DependencyIndexWrapper> at galaxy/scripts/components/admin/Dependencies/DependencyIndexWrapper.vue
                 <ResolutionIndex> at galaxy/scripts/components/admin/Dependencies/ResolutionIndex.vue
                   <Root>
warn @ vue.runtime.esm.js?2b0e:619
logError @ vue.runtime.esm.js?2b0e:1884
globalHandleError @ vue.runtime.esm.js?2b0e:1879
handleError @ vue.runtime.esm.js?2b0e:1839
Vue._render @ vue.runtime.esm.js?2b0e:3544
updateComponent @ vue.runtime.esm.js?2b0e:4060
get @ vue.runtime.esm.js?2b0e:4473
Watcher @ vue.runtime.esm.js?2b0e:4462
mountComponent @ vue.runtime.esm.js?2b0e:4067
Vue.$mount @ vue.runtime.esm.js?2b0e:8409
init @ vue.runtime.esm.js?2b0e:3118
createComponent @ vue.runtime.esm.js?2b0e:5972
createElm @ vue.runtime.esm.js?2b0e:5919
createChildren @ vue.runtime.esm.js?2b0e:6047
createElm @ vue.runtime.esm.js?2b0e:5948
patch @ vue.runtime.esm.js?2b0e:6471
Vue._update @ vue.runtime.esm.js?2b0e:3939
updateComponent @ vue.runtime.esm.js?2b0e:4060
get @ vue.runtime.esm.js?2b0e:4473
Watcher @ vue.runtime.esm.js?2b0e:4462
mountComponent @ vue.runtime.esm.js?2b0e:4067
Vue.$mount @ vue.runtime.esm.js?2b0e:8409
init @ vue.runtime.esm.js?2b0e:3118
createComponent @ vue.runtime.esm.js?2b0e:5972
createElm @ vue.runtime.esm.js?2b0e:5919
createChildren @ vue.runtime.esm.js?2b0e:6047
createElm @ vue.runtime.esm.js?2b0e:5948
patch @ vue.runtime.esm.js?2b0e:6471
Vue._update @ vue.runtime.esm.js?2b0e:3939
updateComponent @ vue.runtime.esm.js?2b0e:4060
get @ vue.runtime.esm.js?2b0e:4473
Watcher @ vue.runtime.esm.js?2b0e:4462
mountComponent @ vue.runtime.esm.js?2b0e:4067
Vue.$mount @ vue.runtime.esm.js?2b0e:8409
init @ vue.runtime.esm.js?2b0e:3118
createComponent @ vue.runtime.esm.js?2b0e:5972
createElm @ vue.runtime.esm.js?2b0e:5919
updateChildren @ vue.runtime.esm.js?2b0e:6210
patchVnode @ vue.runtime.esm.js?2b0e:6313
patch @ vue.runtime.esm.js?2b0e:6476
Vue._update @ vue.runtime.esm.js?2b0e:3942
updateComponent @ vue.runtime.esm.js?2b0e:4060
get @ vue.runtime.esm.js?2b0e:4473
run @ vue.runtime.esm.js?2b0e:4548
flushSchedulerQueue @ vue.runtime.esm.js?2b0e:4304
eval @ vue.runtime.esm.js?2b0e:1980
flushCallbacks @ vue.runtime.esm.js?2b0e:1906
Promise.then (async)
timerFunc @ vue.runtime.esm.js?2b0e:1933
nextTick @ vue.runtime.esm.js?2b0e:1990
queueWatcher @ vue.runtime.esm.js?2b0e:4396
update @ vue.runtime.esm.js?2b0e:4538
notify @ vue.runtime.esm.js?2b0e:730
reactiveSetter @ vue.runtime.esm.js?2b0e:1055
showRowDetails @ DependencyIndexMixin.js?efce:42
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1854
invoker @ vue.runtime.esm.js?2b0e:2179
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1854
Vue.$emit @ vue.runtime.esm.js?2b0e:3882
rowClicked @ mixin-tbody-row.js?c324:164
eval @ mixin-tbody-row.js?c324:95
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1854
invoker @ vue.runtime.esm.js?2b0e:2179
original._wrapper @ vue.runtime.esm.js?2b0e:6911
Show 17 more frames
vue.runtime.esm.js?2b0e:1888 TypeError: Cannot read property 'dependency_resolver' of undefined
    at Proxy.render (ResolutionDetails.vue?e2f2:71)
    at VueComponent.Vue._render (vue.runtime.esm.js?2b0e:3542)
    at VueComponent.updateComponent (vue.runtime.esm.js?2b0e:4060)
    at Watcher.get (vue.runtime.esm.js?2b0e:4473)
    at new Watcher (vue.runtime.esm.js?2b0e:4462)
    at mountComponent (vue.runtime.esm.js?2b0e:4067)
    at VueComponent.Vue.$mount (vue.runtime.esm.js?2b0e:8409)
    at init (vue.runtime.esm.js?2b0e:3118)
    at createComponent (vue.runtime.esm.js?2b0e:5972)
    at createElm (vue.runtime.esm.js?2b0e:5919)
    at createChildren (vue.runtime.esm.js?2b0e:6047)
    at createElm (vue.runtime.esm.js?2b0e:5948)
    at VueComponent.patch [as __patch__] (vue.runtime.esm.js?2b0e:6471)
    at VueComponent.Vue._update (vue.runtime.esm.js?2b0e:3939)
    at VueComponent.updateComponent (vue.runtime.esm.js?2b0e:4060)
    at Watcher.get (vue.runtime.esm.js?2b0e:4473)
    at new Watcher (vue.runtime.esm.js?2b0e:4462)
    at mountComponent (vue.runtime.esm.js?2b0e:4067)
    at VueComponent.Vue.$mount (vue.runtime.esm.js?2b0e:8409)
    at init (vue.runtime.esm.js?2b0e:3118)
    at createComponent (vue.runtime.esm.js?2b0e:5972)
    at createElm (vue.runtime.esm.js?2b0e:5919)
    at createChildren (vue.runtime.esm.js?2b0e:6047)
    at createElm (vue.runtime.esm.js?2b0e:5948)
    at VueComponent.patch [as __patch__] (vue.runtime.esm.js?2b0e:6471)
    at VueComponent.Vue._update (vue.runtime.esm.js?2b0e:3939)
    at VueComponent.updateComponent (vue.runtime.esm.js?2b0e:4060)
    at Watcher.get (vue.runtime.esm.js?2b0e:4473)
    at new Watcher (vue.runtime.esm.js?2b0e:4462)
    at mountComponent (vue.runtime.esm.js?2b0e:4067)
    at VueComponent.Vue.$mount (vue.runtime.esm.js?2b0e:8409)
    at init (vue.runtime.esm.js?2b0e:3118)
    at createComponent (vue.runtime.esm.js?2b0e:5972)
    at createElm (vue.runtime.esm.js?2b0e:5919)
    at updateChildren (vue.runtime.esm.js?2b0e:6210)
    at patchVnode (vue.runtime.esm.js?2b0e:6313)
    at VueComponent.patch [as __patch__] (vue.runtime.esm.js?2b0e:6476)
    at VueComponent.Vue._update (vue.runtime.esm.js?2b0e:3942)
    at VueComponent.updateComponent (vue.runtime.esm.js?2b0e:4060)
    at Watcher.get (vue.runtime.esm.js?2b0e:4473)
    at Watcher.run (vue.runtime.esm.js?2b0e:4548)
    at flushSchedulerQueue (vue.runtime.esm.js?2b0e:4304)
    at Array.eval (vue.runtime.esm.js?2b0e:1980)
    at flushCallbacks (vue.runtime.esm.js?2b0e:1906)
```

jmchilton committed:
Update lib/galaxy/webapps/galaxy/api/container_resolution.py

Co-Authored-By: Marius van den Beek <m.vandenbeek@gmail.com>

jmchilton commented:
Thanks for the details review @mvdbeek and thanks for fixing the performance of the resolution grid.

> I'm totally fine with separating containers and dependencies, you've convinced me. 

Let me make that change then and see what this looks like. I'll keep the mixed resolution mode in as an optional API parameter that is marked as experimental and may go away.

>  One thing though is even with the new caching it takes more than an hour to list the container resolver view. Maybe we can just search for containers on demand ?

I have many thoughts on this... my incremental recommendation would be...

- Merge without worrying more about performance initially. The APIs are important and pretty good, we need a way to access this information and this works just fine as is for small toolboxes (Galaxy flavors, targeted public instances, planemo serving a workflow, etc...). It also shouldn't be a problem for Singularity resolution (?) which usegalaxy.* will likely leverage first. Even for Docker - I don't really feel there is another good way to access this information currently - so an hour is better than nothing. The use case I developed this for is checking cloud instances - we should only need to check this information once in a while for QC purposes and such, it isn't an every day thing.
- Iteration 2: Implement tool ID filtering client-side - it is valuable regardless of performance. Doesn't help performance immediately but when paired with iteration 3...
- Iteration 3: Render the grid with the tool ids and bring in the data on demand if there are more than say 500 tools?
   - Iteration 3 - just load the tool IDs without resolution information.
   - Iteration 3.1 - do resolution without the slow resolvers (just ``mulled`` currently), render the information with a spinner next to it implying more complete information is coming.
   - Iteration 3.2 - use client caching to render saved information the same way as 3.1.

  So render what you can, make sure it is marked as incomplete until a full resolution is done allow filtering immediately to prune the space of things that need to be resolved.
   The problem with rendering the resolution only after you ask for it is that I developed this with quickly summarizing a whole toolbox in mind. Maybe it is the right approach though, I will readily admit my proposal above might look like a denial of service against 

- Iteration 4 (though it can be concurrent with client approaches): Cache the ``mulled`` resolution of a tool id on the backend. We should be doing this anyway for reproducibility I think? I would be much less concerned about the hour loading time if it only happened once per instance - we could resolve on install also to further mitigate this.

Typing this all out though I don't know if Iteration 4 shouldn't be Iteration 1. That could be a big win.

This is a fun project and I really love Vue and think our client infrastructure gives us the tools to do this easily... but none of this is on the project roadmap so I'm not sure it is worth the effort immediately.

Update: To be clear here - I mean a semi-permanent cache using Beaker like we do for citations. This is a very similar problem to citations really.

jmchilton committed:
Refactor mulled resolver for fewer logic paths...

... easier to cache.

jmchilton committed:
Use Beaker cache for mulled resolution.

jmchilton committed:
More aggressively cache mulled resolution failures within request.

jmchilton committed:
Dependency resolution view tool id fixes.

jmchilton committed:
Styling improvements for dependency grids.

jmchilton committed:
Refine resolution grids - don't show actions if loading.

jmchilton committed:
Make container resolution via dependency API optional, off by default.

Makes resolution grid a dependency resolution grid. Deviates from the previous legacy API by not showing containers now - reasons for this are laid out here - https://github.com/galaxyproject/galaxy/pull/7125#issuecomment-539516275.

jmchilton committed:
Cleanup dependency resolution details.

jmchilton committed:
More structured resolution details display.

jmchilton committed:
Display a lot more raw data from server in dependency grids.

jmchilton commented:
I've taken multiple passes at improving the UI - the detailed view is a lot more structured now, better presentation, added information about Conda resolver configuration if resolved that way, and added a little button that display underlying JSON for all the objects displayed - that has even more information than presented in the GUI. I think that level of detail is important for developers or admins that this UI is targeted at.

I've removed containers from the "resolution" grid and renamed it "Dependencies". The API still allows mixed resolution if one just adds "include_containers" as a query parameter. We could therefore create a mixed resolution grid again very quickly - either as a new tab or with a checkbox.

I've implemented the Beaker cache for mulled resolution and a per-request miss cache (so null resolutions aren't repeatedly checked).

jmchilton committed:
Implement mapping container resolver.

jmchilton commented:
I think this ready to go on my end, @mvdbeek anything else you need to see from this PR?

I think this sets up to do some nice fixes as bug fixes during the freeze of the next release if we need to.

mvdbeek commented on lib/galaxy/tool_util/deps/containers.py:
```diff
@@ -181,6 +194,19 @@ def __init__(self, app_info):
         self.container_resolvers = self.__build_container_resolvers(app_info)
+        self.mulled_resolution_cache = None
+        if app_info.mulled_resolution_cache_type:
+            # Do not want to make beaker a required dependency of container resolution, that doesn't make a lot
```
Would you be opposed to making this a top-level import with try / except ImportError ?

jmchilton replied:
How about https://github.com/galaxyproject/galaxy/pull/8741/commits/b6b67f1f474e9d9cdeb07e06176338d493c86fb0 - cleaner imports, cleaner dependencies?

mvdbeek replied:
😍

mvdbeek commented on lib/galaxy/tool_util/deps/mulled/util.py:
```diff
@@ -47,24 +50,56 @@ def quay_repository(namespace, pkg_name):
     The result will be sorted so newest tags are first.
     """
+    if resolution_cache is not None:
+        # Following check is pretty expensive against biocontainers... don't even both doing it
```
```suggestion
        # Following check is pretty expensive against biocontainers... don't even bother doing it
```

jmchilton committed:
Update lib/galaxy/tool_util/deps/mulled/util.py

Co-Authored-By: Marius van den Beek <m.vandenbeek@gmail.com>

jmchilton committed:
Refactor container res. cache injection for clean dependencies.

mvdbeek commented:
Arg, there's a conflict here, if you resolve I'll merge

jmchilton commented:
That inconsiderate jerk @jmchilton is constantly inconveniencing me! Rebased.

mvdbeek commented:
Tests are clearly failing due to wheels issues.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -766,6 +766,30 @@ mapping:
+
+      mulled_resolution_cache_data_dir:
+        type: str
+        default: citations/data
```
This seem a copy/paste mistake.

nsoranzo commented on lib/galaxy/webapps/galaxy/config_schema.yml:
```diff
@@ -766,6 +766,30 @@ mapping:
+
+      mulled_resolution_cache_lock_dir:
+        type: str
+        default: citations/locks
```
Here too.

nsoranzo reviewed (commented):
@jmchilton Can you please fix the 2 defaults and update config files with `make config-rebuild` ?
